### ISSUE TITLE
Increase multithreading processing capability

### ISSUE THREAD

**Issue by chenziliang0725** [UB-Mannheim/zotero-ocr#72]
Tesseract and poppler only pureduce pages one by one now, when there are dozens of pages, it work slowly.
can we increase multithreading processing capability

**Comment by aborel** [comment]
There could be ways to do this, at least for tesseract. I'll take a look.

**Comment by stweil** [comment]
Currently the code runs the Tesseract executable with a list of page images. Then Tesseract processes those images one by one which takes some time.

zotero-ocr could accelerate the recognition by running several parallel Tesseract processes, but that would increase the complexity because it would require an additional processing step to combine the results of the different Tesseract processes.

I think it would be easier to add a reasonable multithreading to the Tesseract code. The current multithreading in Tesseract is not helpful, but multithreading on the page level would have a large benefit.

**Comment by aborel** [comment]
I agree. My plan was to investigate the current Tesseract situation before writing any code here, so thanks for this input.

**Comment by aborel** [comment]
The way I see it, the ideal situation for us would be if someone implemented this Tesseract issue https://github.com/tesseract-ocr/tesseract/issues/3750 . Then we'd get the functionality at a minimal cost (maybe just adding a preference for the number of threads).

However, the Tesseract issue is 2 years old with no activity in sight, so I'm not really confident it will happen soon. I don't have the proper skill set to contribute on that side, unfortunately. However, I think the added complexity to implement this within the Zotero-OCR code might be manageable... I'd like to try.

**Comment by aborel** [comment]
I think we could use Poppler's pdfunite to re-assemble multiple PDFs https://manpages.debian.org/testing/poppler-utils/pdfunite.1.en.html . We could probably rely on the pdftoppm settings to locate the executable.

Of course we also must take care of the other output files (text, hocr...) but it looks feasible.

**Comment by stweil** [comment]
There could be a fallback to the current processing if no pdfunite was found. Would this be sufficient? Then I'll ask my AI assistant to implement parallel processing of pages. How many pages should be processed in parallel? Up to 50% load? Or 100% load? Or use a new setting (number of parallel tesseract jobs)?

**Comment by stweil** [comment]
My assistant now implemented parallel processing with up to 4 pages processed in parallel. So it accelerates pdf.js and tesseract by a factor of 4, the total time for a multipage document will shrink to ~25% (or a little bit more, because the final step must merge the results in a single text, hOCR or PDF file). Pull request #135 introduces pdf.js, so I can send a pull request for the issue here after #135 was merged.

**Comment by aborel** [comment]
Perhaps we could evaluate the feasibility of a switch to https://tesseract.projectnaptha.com/ (or some functional equivalent) before we revisit this issue? I'd be happy to remove that important external tool dependency, and there is a chance that executing all processes with pure Javascript function calls might make the parallel code easier to follow and maintain. What is your opinion about this?

**Comment by stweil** [comment]
Yes, running the Javascript of Tesseract would make the installation easier, but also cost performance. But I think that this can be addressed later. I propose to start with PR #135, then #136, then I can send a PR for parallel processing with the current Tesseract. After these changes, issue #125 might be addressed.

Then we can close half of the open issues, and if time permits, we can try using the JavaScript port of Tesseract.

**Comment by aborel** [comment]
Performance can be an issue in theory, but do we really have evidence for that? (honest question, I haven't really checked, maybe you did)

**Comment by stweil** [comment]
I have never tried the JavaScript port. Therefore I don't know how much slower JavaScript code is in comparision to optimized C++ code.

**Comment by aborel** [comment]
I've seen some quite efficient scientific computing applications written in 100% native Javascript, executing in the user's browser - the performance overhead could be less than you imagine. Definitely worth investigating! This being said, if you prefer to close some issues even if it means re-factoring code later, I can live with that. Just try to prompt your agent to produce code that will be as future-proof as possible ;-)
